src/coreclr/jit/optimizer.cpp

```diff
@@ -4798,124 +4866,175 @@ void Compiler::optMarkLoopHeads()
 }
 
 //-----------------------------------------------------------------------------
-// optFindLoops: find loops in the function.
-//
-// The JIT recognizes two types of loops in a function: natural loops and "general" (or "unnatural") loops.
-// Natural loops are those which get added to the loop table. Most downstream optimizations require
-// using natural loops. See `optFindNaturalLoops` for a definition of the criteria for recognizing a natural loop.
-// A general loop is defined as a lexical (program order) range of blocks where a later block branches to an
-// earlier block (that is, there is a back edge in the flow graph), and the later block is reachable from the earlier
-// block. General loops are used for weighting flow graph blocks (when there is no block profile data), as well as
-// for determining if we require fully interruptible GC information.
-//
-// Notes:
-//  Also (re)sets all non-IBC block weights, and marks loops potentially needing alignment padding.
+// optResetLoopInfo: reset all loop info in preparation for rebuilding the loop table, or preventing
+// future phases from accessing loop-related data.
 //
-PhaseStatus Compiler::optFindLoops()
+void Compiler::optResetLoopInfo()
 {
-    noway_assert(opts.OptimizationEnabled());
-    assert(fgDomsComputed);
-
 #ifdef DEBUG
     if (verbose)
     {
-        printf("*************** In optFindLoops()\n");
+        printf("*************** In optResetLoopInfo()\n");
     }
 #endif
 
-    optMarkLoopHeads();
-
-    optSetBlockWeights();
+    optLoopCount        = 0;       // This will force the table to be rebuilt
+    optLoopTable        = nullptr; // This will cause users to crash if they use the table when it is considered empty.
+    loopAlignCandidates = 0;
```

Comment:
Should we have a loop table epoch, in case loop numbers get recorded in random places and can become stale?

Reply:
I like the "epoch" idea, and I'll add the basics, but I don't know how/where it will get validated. Will keep that part in mind.